### PR TITLE
Allow drop-ins for transient units

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Review by yuwata — APPROVED** [review]
LGTM

**Review by bluca — APPROVED** [review]

File: src/core/unit.c [hunk 1]
```diff
@@ -4853,7 +4853,6 @@ bool unit_is_pristine(Unit *u) {
         return IN_SET(u->load_state, UNIT_NOT_FOUND, UNIT_LOADED) &&
                !u->fragment_path &&
                !u->source_path &&
-               strv_isempty(u->dropin_paths) &&
                !u->job &&
                !u->merged_into;
```
Comment (poettering) [inline src/core/unit.c:4860]: can you add a comment here, mentioning that we explicitly do *not* check for drop-in paths, and why?

**Comment by poettering** [comment]
Hmm, so thinking about this, I think this either won't work as intended or should carry a test that ensures it works as intended...

Here's the problem I see with this: if a unit has a dropin already by the time we want to make it a transient unit, then its dropins are already *loaded*, not just tracked.

So when we then deserialize the dbus transient unit props into it, they will then override the dropins we already loaded. And we won't load them again then, afaics.

Hence, I am pretty sure this doesn't actually work as intended, as the dropins for transient units would then be overriden by the transient settings initially, but once you use "systemctl daemon-reload" they would then suddenly override the transient settings, and that's certainly wrong. i.e. the order of overriding would be reversed before and after the first daemon-reload after the transient unit is created.

**Comment by poettering** [comment]
I don't really know how to address this properly. The thing is: it should be OK if transient units are already referenced (and thus exist in memory) by the time they are invoked via StartTransientUnit(). But if they exist in memory then they should also have all existing info loaded (as we have to load them to figure out if they exist), hence it is *correct* to to already load the drop-ins. But if we load the drop-ins then we are fucked with making the unit into a transient one, since the transient unit definition should always the the "lowest" layer in the stack of configuration read for the unit.

**Comment by poettering** [comment]
(people would probably build you a statue for adding pinpointed reloads though)

**Comment by keszybz** [comment]
> Hmm, so thinking about this, I think this either won't work as intended or should carry a test that ensures it works as intended...

I took the second path ;) A fairly extensive test is added in TEST-15-DROPINS. This took longer than I expected because to make it nice and clean the existing code needed to be refactored a bit. I adjusted the comment as requested too, no other changes.

**Comment by keszybz** [comment]
Re: pinpoint reloads. I agree we should add this. With the rework you did to have dependency atoms, I think we're 90% of the way to have this working. I'll put it on my TODO list, but I think we need to get the release out first before embarking on new stuff…

**Review by bluca — APPROVED** [review]

**Comment by poettering** [comment]
So, what happens if I do StandardInputText=foo inside a drop-in? Each time the line is used we'll add to the input buffer. So if i have a drop-in that has that, and I also set StandardInputText=bar from the transient unit definition in the dbus message, will I end up with "foo\nbar\nfoo\n" in standard input?

That is wrong though, it should just be "bar\nfoo".

How do you intend to address without ding the full unit reset logic that pinpointed reload requires?

**Comment by poettering** [comment]
> Re: pinpoint reloads. I agree we should add this. With the rework you did to have dependency atoms, I think we're 90% of the way to have this working. I'll put it on my TODO list, but I think we need to get the release out first before embarking on new stuff…

Dunno, this is a new feature, there's no need to hurry this in the current release. This is 253 material I am sure.

**Comment by keszybz** [comment]
I added a test with StandardInputData. The service doesn't actually use the data, but this shouldn't matter. `systemctl show -P` shows that we get the expected contents. PTAL.

**Comment by poettering** [comment]
hmm? so how come the stuff is not read twice? I really don't follow how this could possibly work.

**Comment by keszybz** [comment]
There are two cases:
- a service (i.e. a unit that needs a fragment). In this case, loading of the unit with drop-ins before the fragment is written actually fails, thus we only load the drop-ins after the main fragment, and everything is fine.
- a slice (i.e. a unit that is fine without a fragment). In this case, I think we indeed load the drop-ins first, then load the main fragment, and then the drop-ins again. This would seem wrong, but it doesn't really matter because it is hard to come up with a case where the doubled loading would cause a visible difference:
     - most settings are "single-valued", i.e. repeated configuration cleanly overwrites earlier configuration
     - in case of "accumulating" settings like lists of things, usually the order of configuration doesn't matter and repeated values are coalesced (Wants=, Requires=, After=, AllowedCPUs=, etc, etc)
     - there are some settings where repeated values matter, e.g. StandardInputData=. But afaict, none of those settings can be applied to slices or devices.

Thus… I think we should merge this and get the release out. There are some corner cases which are a bit messy, but they are very much corner cases that anyone is unlikely to care about.

**Comment by poettering** [comment]
What about the other transient unit types, i.e. mounts and sockets and timers and automounts and paths and scopes.

What about Documentation=? afaics it doesn't remove duplicates? And ConditionFileName= and the lot doesn't either, I think?

This seems very incomplete, and I am not sure why we need to hurry this? THis is not a regression?

**Comment by keszybz** [comment]
> What about the other transient unit types, i.e. mounts and sockets and timers and automounts and paths and scopes.

As described above, the only problem is with fragment-less types: device, slice, scope. All of those generally don't have too many settings.

> What about Documentation=? afaics it doesn't remove duplicates? And ConditionFileName= and the lot doesn't either, I think?

Documentation doesn't. But I think we should de-duplicate Documentation, that'd seem generally useful. I think we also don't deduplicate conditions.

There is an issue here. I'm not trying to deny that. Nevertheless, I think it's something that is not a problem in practice. It seems rather unlikely that someone would set Documentation or Condition* via a slice / device / scope drop-in, and even if they do, the result will be just some repeats or unexpected ordering.
 
> This seems very incomplete, and I am not sure why we need to hurry this? THis is not a regression?

Yes, this is a very significant regression. We added support for "hierarchical drop-ins" some time ago, and when people started using them more widely, it turned out that this breaks transient units. We didn't see it before, because we didn't have many such drop-ins. But right now they are "incompatible" and that is just bad.

**Review by bluca — APPROVED** [review]

**Comment by keszybz** [comment]
I'll push this to Fedora for now. I think we should add this to the stable branches too.

**Comment by keszybz** [comment]
Queued up for v251.8.

**Pull request merged**